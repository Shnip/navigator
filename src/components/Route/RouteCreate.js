import React from 'react';
import PropTypes from 'prop-types';
import Route from './routeFactory';
import {NotificationContainer, NotificationManager} from 'react-notifications';


class RouteCreate extends React.Component {
  state = {
    route: ''
  }

  componentDidMount() {
    this.createYaSearch()
  }

  createYaSearch = () => {
    const ymaps = window.ymaps;

    if (!ymaps) return false;

    ymaps.ready(() => {
      const suggestView = new ymaps.SuggestView(
        'route__input',
        {
            results: '5',
            provider: 'yandex#map',
        }
      );

      suggestView.events.add('select', (e) => {
        const selected = e.get('item').value;
        if (this.state.route !== selected) {
          this.setState({route: selected});
        }
      })
    })
  }

  onSubmit = async (e) => {
    e.preventDefault();

    let {route} = this.state;
    if (!route.trim()) return this.createNotification('error', 'Route is empty');

    let points = [];

    await this.geocode(route)
              .then(res => points = res.geoObjects.toArray())
              .catch(e => console.log(e))

    if (points.length === 0) return this.createNotification('error', 'Sorry, we can\'t find the address');

    const isCoordinates = route.match(/\d+\.?\d*,?\s?\d+\.?\d*/gi);
    if (isCoordinates) route = points[0].properties.get('text');

    const geoObject = Object.assign({},
      {coordinates: points[0].geometry.getCoordinates()},
      {balloon: points[0].properties.get('balloonContent')}
    );

    this.props.addRoute(new Route(route, geoObject));
    this.setState({route: ''});
  }

  geocode = (route) => {
    const ymaps = window.ymaps;
    if (!ymaps) return false

    return ymaps.geocode(route, {
      json: false,
      provider: 'yandex#map',
      results: 1
    })
  }

  createNotification = (type, message) => {
    switch (type) {
      case 'error':
        NotificationManager.error('Sorry, we can\'t find the address', 'Error', 6000);
        break;
      default:
        break;
    }
  }

  render() {
    return (
      <>
        <NotificationContainer/>
        <div className="routes__create">
          <h2 className="title title--white routes__title">
            <span className="text-bold">Create</span>
            &nbsp;
            <span className="text-thin">Route</span>
          </h2>
          <form name="routes__form" className="form routes__form" onSubmit={this.onSubmit}>
            <div className="form__input-group">
              <input type="text"
                     id="route__input"
                     className="form__input"
                     value={this.state.route}
                     placeholder="Москва, метро Арбатская или 55.753595 37.621031"
                     onChange={e => this.setState({route: e.target.value})}
                     />
              <button className="btn btn--light-theme">Add</button>
            </div>
          </form>
        </div>
      </>
    );
  }
}

RouteCreate.propTypes = {
  addRoute: PropTypes.func.isRequired
}

export default RouteCreate;
