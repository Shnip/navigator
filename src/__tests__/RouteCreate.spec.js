import React from 'react';
import { shallow, mount } from 'enzyme';
import RouteCreate from '../components/Route/RouteCreate';

describe('test ReactCreate component', () => {
  const mockAddRoute = jest.fn();
  const props = {
    addRoute: mockAddRoute
  }

  const initialState = {
    route: ''
  }

  const mockYmaps = {
    geocode: (route) => {
      return new Promise((res, rej) => {
        const rightResult = {
          geoObjects: {
            toArray: () => [
              {
                properties: { get: text => 'Russia, Moscow' },
                geometry: {getCoordinates: () => [54.654, 65.875]}
              }
            ]
          }
        }
        const wrongResult = {
          geoObjects: { toArray: () => [] }
        }

        if (route === 'Moscow') res(rightResult);
        res(wrongResult)
      })
    },
    ready: () => {}
  }
  window.ymaps = mockYmaps;

  const component = shallow(<RouteCreate {...props} />);

  it('renders components', () => {
    expect(component.find('.form__input')).toHaveLength(1);
    expect(component.find('.btn')).toHaveLength(1);
  });

  it('renders NotificationContainer', () => {
    expect(component.find('NotificationContainer')).toHaveLength(1)
  })

  it('should change state via input', () => {
    component.find('#route__input').simulate('change', {
      target: {
        value: 'Moscow'
      }
    });

    expect(component.state().route).toEqual('Moscow');
  });

  it('shouldnt call addRoute when route is empty string', () => {
    component.find('.routes__form').simulate('submit', {
      preventDefault: jest.fn()
    })

    setTimeout(() => expect(mockAddRoute).toHaveBeenCalledTimes(0), 4000);
  });

  it('should call addRoute when route isn\'t empty', () => {
    component.setState({ route: 'Moscow'});
    component.find('.routes__form').simulate('submit', {
      preventDefault: () => {}
    })

    setTimeout(() => expect(mockAddRoute).toHaveBeenCalledTimes(1), 4000);
  });

  it('shouldn\'t call addRoute when route isn\'t found', () => {
    component.setState({ route: 'qwerty'});
    component.find('.routes__form').simulate('submit', {
      preventDefault: () => {}
    })

    setTimeout(() => expect(mockAddRoute).toHaveBeenCalledTimes(0), 4000);
  });

  it ('renders properly', () => {
    expect(component).toMatchSnapshot();
  });

  afterEach(() => {
    component.setState(initialState);
    mockAddRoute.mockClear();
  });
})
