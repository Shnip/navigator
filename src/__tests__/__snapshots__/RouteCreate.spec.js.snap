// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ReactCreate component renders properly 1`] = `ShallowWrapper {}`;
